src/Common/Config/ConfigProcessor.cpp

```diff
@@ -622,6 +627,52 @@ ConfigProcessor::Files ConfigProcessor::getConfigMergeFiles(const std::string &
     return files;
 }
 
+XMLDocumentPtr ConfigProcessor::parseConfig(const std::string & config_path)
+{
+    fs::path p(config_path);
+    std::string extension = p.extension();
+    boost::algorithm::to_lower(extension);
+
+    if (extension == ".xml")
+        return dom_parser.parse(config_path);
+    else if (extension == ".yaml" || extension == ".yml")
+        return YAMLParser::parse(config_path);
+    else
+    {
+        /// Suppose non regular file parsed as XML, such as pipe: /dev/fd/X (regardless it has .xml extension or not)
+        if (!fs::is_regular_file(config_path))
+            return dom_parser.parse(config_path);
+
+        /// If the regular file begins with < it might be XML, otherwise it might be YAML.
+        bool maybe_xml = false;
+        {
+            std::ifstream file(config_path);
+            if (!file.is_open())
+                throw Exception(ErrorCodes::CANNOT_LOAD_CONFIG, "Unknown format of '{}' config", config_path);
+
+            std::string line;
+            while (std::getline(file, line))
+            {
+                const size_t pos = firstNonWhitespacePos(line);
+                if (pos == std::string::npos)
+                    continue;
+
+                if (pos < line.size() && '<' == line[pos])
+                {
+                    maybe_xml = true;
+                }
+
+                break;
+            }
+        }
+
+        if (maybe_xml)
+            return dom_parser.parse(config_path);
+        else
+            return YAMLParser::parse(config_path);
```

Comment:
```suggestion
            while (std::getline(file, line))
            {
                const size_t pos = firstNonWhitespacePos(line);

                if (pos < line.size() && '<' == line[pos])
                    return dom_parser.parse(config_path);
            }
        }

        return YAMLParser::parse(config_path);
```

Reply:
Hi @yariks5s, thanks a lot for your reviews! The suggested code is more concise. I have some small considerations:
1. If the configuration file format is YAML, it will traverse each line of the file will cost more time since it is not possible to exit the traversal early.
2. `std::ifstream file(config_path)` already opened the file (not closed), `dom_parser.parse(config_path)` will open the same file and read again, since they are all read operations, there is no need to close and then open it, right?

Reply:
Considering your message the possible solution can be 
``` C++
            while (std::getline(file, line))
            {
                const size_t pos = firstNonWhitespacePos(line);

                if (pos < line.size() && '<' == line[pos])
                    maybe_xml = true;
                else if (pos != std::string::npos)
                    break;
            }
        }
        if (maybe_xml)
                return dom_parser.parse(config_path);
        return YAMLParser::parse(config_path);
```

Reply:
`maybe_xml = true;` also needs to be followed by a break?

Reply:
yes